repository/src/lib.rs

```diff
@@ -196,8 +196,88 @@ impl<T: RawRepository> DistributedRepository<T> {
+                "block commit is not a descendant of the current finalized branch"
+            ));
+        }
+
+        // Construct a commit list starting from the next commit of the last finalized block to the `branch_tip_commit`(the most recent commit of the branch)
+        let ancestor_commits = self.raw.list_ancestors(*block_commit, Some(256)).await?;
+        let position = ancestor_commits
+            .iter()
+            .position(|c| *c == current_finalized_commit)
+            .expect("TODO: handle the case where it exceeds the limit");
+        let new_semantic_commits = stream::iter(
+            ancestor_commits
+                .iter()
+                .take(position)
+                .rev()
+                .cloned()
+                .map(|c| {
+                    let raw = &self.raw;
+                    async move { raw.read_semantic_commit(c).await.map(|x| (x, c)) }
+                }),
+        )
+        .buffered(256)
+        .collect::<Vec<_>>()
+        .await;
+        let mut new_semantic_commits = new_semantic_commits
+            .into_iter()
+            .collect::<Result<Vec<_>, _>>()?;
+        // Add most recent commit of the branch to the list since it is not included in the ancestor commits
+        let block_semantic_commit = self.raw.read_semantic_commit(*block_commit).await?;
+        new_semantic_commits.push((block_semantic_commit, *block_commit));
+        let new_commits = new_semantic_commits
+            .into_iter()
+            .map(|(commit, hash)| {
+                from_semantic_commit(commit)
+                    .map_err(|e| (e, hash))
+                    .map(|x| (x, hash))
+            })
+            .collect::<Result<Vec<_>, _>>()
+            .map_err(|(error, hash)| anyhow!("failed to convert the commit {}: {}", hash, error))?;
+
+        let start_header = self.get_last_finalized_block_header().await?;
+        let reserved_state = self.get_reserved_state().await?;
+
+        // Verify all the new commits
+        let mut verifier =
+            CommitSequenceVerifier::new(start_header.clone(), reserved_state.clone())
+                .map_err(|e| anyhow!("failed to create a verifier: {}", e))?;
+        for (new_commit, _) in &new_commits {
+            if verifier.apply_commit(new_commit).is_err() {
+                return Err(anyhow!("CSV failed"));
+            }
+        }
+
+        // Verify finalization proof, not changing the fp branch
+        // This is the difference with 'finalize' method
+        let fp_commit_hash = self.raw.locate_branch(FP_BRANCH_NAME.into()).await?;
+        let fp_semantic_commit = self.raw.read_semantic_commit(fp_commit_hash).await?;
+        let finalization_proof: FinalizationProof =
+            serde_json::from_str(&fp_semantic_commit.body).unwrap();
+        let semantic_commit = self.raw.read_semantic_commit(*block_commit).await?;
+        let commit = format::from_semantic_commit(semantic_commit).map_err(|e| anyhow!(e))?;
```

Comment:
```suggestion
        let block_commit = format::from_semantic_commit(block_semantic_commit).map_err(|e| anyhow!(e))?;
```
you will need to clone line 238 since move occurs there

Reply:
Please check my new version. This one is outdated.